src/main/kotlin/biokotlin/genome/MAFProcessingUtils.kt

```diff
@@ -447,4 +452,321 @@ fun getCoverageIdentityPercentForMAF(mafFile:String, region:String = "all"): Dat
+    val chromToRanges = mutableMapOf<String, RangeMap<Int, MAFRecord>>() // needed for pulling sequence from submap
+    for (mafrec in source) {
+        val chrom = mafrec.refRecord.chromName
+        // this needed for creating the list of gap ranges
+        var chrMAFRecList = chromToRangesSRC[chrom]
+        if (chrMAFRecList == null) {
+            chrMAFRecList = ArrayList<MAFRecord>()
+            chromToRangesSRC[chrom] =  chrMAFRecList
+        }
+        chrMAFRecList!!.add( mafrec)
+
+        // create a map of chromosome name -> RangeMap
+        // This one needed for getting subranges later from the gaps
+        var chrMap = chromToRanges[chrom]
+        if (chrMap == null) {
+            chrMap = TreeRangeMap.create()
+            chromToRanges[chrom] = chrMap
+        }
+        val start = mafrec.refRecord.start
+        val end = mafrec.refRecord.start + mafrec.refRecord.size - 1
+        chrMap!!.put(Range.closed(start,end), mafrec)
+    }
+
+    // GEt the target positions
+    // This creates a map of chromName (key) to a list of MAFtoGVCFPlugin.MAFRecords (value)
+    val chromToRangesTGT = target.groupBy{it.refRecord.chromName}
+
+    val addedBlocks = mutableListOf <MAFRecord>()
+    // Not all chromosomes will be in both target and src
+    println("augmentList: number of chromosomes in chromToMAFRecords = ${chromToRangesSRC.keys.size}")
+    for (chrom in chromToRangesSRC.keys) {
+        val srcRanges = chromToRangesSRC[chrom]
+        val tgtRanges = chromToRangesTGT[chrom]
+
+        if (tgtRanges == null) {
+            addedBlocks.addAll(srcRanges!!.toList())
+            println("augmentList: NO target ranges for chrom ${chrom} - adding all ranges from source chrom ${chrom} to target's list")
+            continue
+        }
+
+        val newRanges = findGaps(tgtRanges!!,srcRanges!!)
+        if (newRanges == null) {
+            println("augmentList: newRanges for chrom ${chrom} is NULL ,continue")
+            continue
+        }
+
+        println("augmentList: found gaps: process gaps for chrom ${chrom}, number of gaps=${newRanges.size}")
+        // take those ranges, run through Peter's code to get sequence
+        for (gap in newRanges) {
+
+            val sourceRangeMap = chromToRanges[chrom]
+            if (sourceRangeMap != null) {
+                val gapRecords = sourceRangeMap.subRangeMap(gap).asMapOfRanges()
+                if (gapRecords.size > 0) {
+                    for (sourceRecord in gapRecords.values) {
+                        val subRecord = extractSubMafRecord(gap.lowerEndpoint(),gap.upperEndpoint(), sourceRecord)
+                        if (subRecord != null) addedBlocks.add(subRecord)
+                    }
+                }
+            }
+        }
+    }
+
+    println("augmentList at end: size of addedBlocks ${addedBlocks.size}")
+    target.addAll(addedBlocks)
+}
+
 
+fun extractSubMafRecord(start: Int, end: Int, mafRecord: MAFRecord) : MAFRecord? {
```

Comment:
Add a comment to the header saying what this method is trying to achieve.

Reply:
added